Mathlib/CategoryTheory/Generator.lean

```diff
@@ -567,4 +569,186 @@ theorem wellPowered_of_isDetector [HasPullbacks C] (G : C) (hG : IsDetector G) :
   haveI := small_subsingleton ({G} : Set C)
   wellPowered_of_isDetecting hG
 
+theorem wellPowered_of_isSeparator [HasPullbacks C] [Balanced C] (G : C) (hG : IsSeparator G) :
+    WellPowered C := wellPowered_of_isDetecting hG.isDetector
+
+section HasGenerator
+
+section Definitions
+
+variable (C)
+
+/--
+For a category `C` and an object `G : C`, `G` is a separator of `C` if
+the functor `C(G, -)` is faithful.
+
+While `IsSeparator G : Prop` is the proposition that `G` is a separator of `C`,
+an `HasSeparator C : Prop` is the proposition that such a separator exists.
+Note that `HasSeparator C` is a proposition. It does not designate a favored separator
+and merely asserts the existence of one.
+-/
+class HasSeparator : Prop where
+  hasSeparator : ∃ G : C, IsSeparator G
+
+/--
+For a category `C` and an object `G : C`, `G` is a coseparator of `C` if
+the functor `C(-, G)` is faithful.
+
+While `IsCoseparator G : Prop` is the proposition that `G` is a coseparator of `C`,
+an `HasCoseparator C : Prop` is the proposition that such a coseparator exists.
+Note that `HasCoseparator C` is a proposition. It does not designate a favored coseparator
+and merely asserts the existence of one.
+-/
+class HasCoseparator : Prop where
+  hasCoseparator : ∃ G : C, IsCoseparator G
+
+/--
+For a category `C` and an object `G : C`, `G` is a detector of `C` if
+the functor `C(G, -)` reflects isomorphisms.
+
+While `IsDetector G : Prop` is the proposition that `G` is a detector of `C`,
+an `HasDetector C : Prop` is the proposition that such a detector exists.
+Note that `HasDetector C` is a proposition. It does not designate a favored detector
+and merely asserts the existence of one.
+-/
+class HasDetector : Prop where
+  hasDetector : ∃ G : C, IsDetector G
+
+/--
+For a category `C` and an object `G : C`, `G` is a codetector of `C` if
+the functor `C(-, G)` reflects isomorphisms.
+
+While `IsCodetector G : Prop` is the proposition that `G` is a codetector of `C`,
+an `HasCodetector C : Prop` is the proposition that such a codetector exists.
+Note that `HasCodetector C` is a proposition. It does not designate a favored codetector
+and merely asserts the existence of one.
+-/
+class HasCodetector : Prop where
+  hasCodetector : ∃ G : C, IsCodetector G
+
+end Definitions
+
+section Choice
+
+variable (C)
+
+/--
+Given a category `C` that has a separator (`HasSeparator C`), `separator` is an arbitrarily
+chosen separator of `C`.
+-/
+noncomputable def separator [HasSeparator C] : C :=
+  Classical.indefiniteDescription _ HasSeparator.hasSeparator |>.1
```

Comment:
Sorry, I think I told you a suboptimal way to write this and the idiomatic way is
```suggestion
  HasSeparator.hasSeparator.choose
```
and similarly `HasSeparator.hasSeparator.choose_spec` for the specification theorem below.